File: VB/FilteringBehavior.vb
Imports DevExpress.Data.Filtering
Imports DevExpress.Data.Filtering.Helpers
Imports DevExpress.Mvvm.UI.Interactivity
Imports DevExpress.Xpf.Grid
Imports System
Imports System.Collections
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Data
Imports System.Linq

Namespace DXGridSample
    Public Class FilteringBehavior
        Inherits Behavior(Of TableView)

        Protected ReadOnly Property Grid() As GridControl
            Get
                Return AssociatedObject.Grid
            End Get
        End Property
        Protected ReadOnly Property View() As TableView
            Get
                Return AssociatedObject
            End Get
        End Property
        Protected ReadOnly Property ItemsSource() As IEnumerable(Of Object)
            Get
                If TypeOf Grid.ItemsSource Is DataTable Then
                    Return CType(Grid.ItemsSource, DataTable).DefaultView.Cast(Of Object)()
                ElseIf TypeOf Grid.ItemsSource Is IEnumerable Then
                    Return DirectCast(Grid.ItemsSource, IEnumerable).Cast(Of Object)()
                End If
                Return Nothing
            End Get
        End Property
        Protected ReadOnly Property DataType() As Type
            Get
                Return ItemsSource.ElementAt(0).GetType()
            End Get
        End Property

        Protected Overrides Sub OnAttached()
            MyBase.OnAttached()
            AddHandler View.ShowFilterPopup, AddressOf ShowFilterPopup
        End Sub
        Protected Overrides Sub OnDetaching()
            RemoveHandler View.ShowFilterPopup, AddressOf ShowFilterPopup
            MyBase.OnDetaching()
        End Sub

        Protected Overridable Sub ShowFilterPopup(ByVal sender As Object, ByVal e As FilterPopupEventArgs)
            If e.Column.FilterPopupMode <> FilterPopupMode.CheckedList OrElse ItemsSource Is Nothing OrElse ItemsSource.Count() = 0 Then
                Return
            End If
            Dim filter = RemoveColumn(Grid.FilterCriteria, e.Column.FieldName)
            If ReferenceEquals(filter, Nothing) Then
                Return
            End If
            e.ComboBoxEdit.ItemsSource = GetFilterItems(filter, e.Column)
        End Sub
        Protected Overridable Function GetFilterItems(ByVal filter As CriteriaOperator, ByVal column As ColumnBase) As Object
            Dim properties = TypeDescriptor.GetProperties(ItemsSource.ElementAt(0))
            Dim info = properties.Cast(Of PropertyDescriptor)().First(Function(x) x.Name = column.FieldName)
            Dim expressionEvaluator = New ExpressionEvaluator(properties, filter)
            Dim distinctValues = expressionEvaluator.Filter(ItemsSource).Cast(Of Object)().Select(Function(obj) info.GetValue(obj)).Distinct()
            Return distinctValues.Select(Function(x) New CustomComboBoxItem With { _
                .EditValue = x, _
                .DisplayValue = column.ActualEditSettings.GetDisplayTextFromEditor(x) _
            }).ToList()
        End Function

        Public Shared Function RemoveColumn(ByVal filter As CriteriaOperator, ByVal fieldName As String) As CriteriaOperator
            If ReferenceEquals(filter, Nothing) Then
                Return Nothing
            End If
            Dim splitted As IDictionary(Of OperandProperty, CriteriaOperator) = CriteriaColumnAffinityResolver.SplitByColumns(filter)
            splitted.Remove(New OperandProperty(fieldName))
            Return GroupOperator.And(splitted.Values)
        End Function
    End Class
End Namespace

File: CS/FilteringBehavior.cs
using DevExpress.Data.Filtering;
using DevExpress.Data.Filtering.Helpers;
using DevExpress.Mvvm.UI.Interactivity;
using DevExpress.Xpf.Grid;
using System;
using System.Collections;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Linq;

namespace DXGridSample {
    public class FilteringBehavior : Behavior<TableView> {
        protected GridControl Grid { get { return AssociatedObject.Grid; } }
        protected TableView View { get { return AssociatedObject; } }
        protected IEnumerable<object> ItemsSource
        {
            get
            {
                if (Grid.ItemsSource is DataTable)
                    return ((DataTable)Grid.ItemsSource).DefaultView.Cast<object>();
                else if (Grid.ItemsSource is IEnumerable)
                    return ((IEnumerable)Grid.ItemsSource).Cast<object>();
                return null;
            }
        }
        protected Type DataType { get { return ItemsSource.ElementAt(0).GetType(); } }

        protected override void OnAttached() {
            base.OnAttached();
            View.ShowFilterPopup += ShowFilterPopup;
        }
        protected override void OnDetaching() {
            View.ShowFilterPopup -= ShowFilterPopup;
            base.OnDetaching();
        }

        protected virtual void ShowFilterPopup(object sender, FilterPopupEventArgs e) {
            if (e.Column.FilterPopupMode != FilterPopupMode.CheckedList || ItemsSource == null || ItemsSource.Count() == 0)
                return;
            var filter = RemoveColumn(Grid.FilterCriteria, e.Column.FieldName);
            if (ReferenceEquals(filter, null))
                return;
            e.ComboBoxEdit.ItemsSource = GetFilterItems(filter, e.Column);
        }
        protected virtual object GetFilterItems(CriteriaOperator filter, ColumnBase column) {
            var properties = TypeDescriptor.GetProperties(ItemsSource.ElementAt(0));
            var info = properties.Cast<PropertyDescriptor>().First(x => x.Name == column.FieldName);
            var expressionEvaluator = new ExpressionEvaluator(properties, filter);
            var distinctValues = expressionEvaluator.Filter(ItemsSource).Cast<object>().Select(obj => info.GetValue(obj)).Distinct();
            return distinctValues.Select(x => new CustomComboBoxItem { EditValue = x, DisplayValue = column.ActualEditSettings.GetDisplayTextFromEditor(x) }).ToList();
        }

        public static CriteriaOperator RemoveColumn(CriteriaOperator filter, string fieldName) {
            if (ReferenceEquals(filter, null))
                return null;
            IDictionary<OperandProperty, CriteriaOperator> splitted = CriteriaColumnAffinityResolver.SplitByColumns(filter);
            splitted.Remove(new OperandProperty(fieldName));
            return GroupOperator.And(splitted.Values);
        }
    }
}